1.process_data/nbconverted/single_cell_qc.py

```diff
@@ -0,0 +1,277 @@
+# 
+# - `Compactness`, which detects irregular nuclei and nuclei containing holes.
+
+# ### Detect outliers and show single-cell image crops with CytoDataFrame (cdf)
+
+# In[9]:
+
+
+# find mis-segmented nuclei due using thresholds that maximizes
+# removing most technical outliers and minimizes good cells
+feature_thresholds = {
+    "Nuclei_AreaShape_Compactness": 4.05,
+}
+
+poor_nuclei_outliers = cosmicqc.find_outliers(
+    df=df_merged_single_cells,
+    metadata_columns=metadata_cols,
+    feature_thresholds=feature_thresholds
+)
+
+poor_nuclei_outliers_cdf = CytoDataFrame(
+    data=pd.DataFrame(poor_nuclei_outliers),
+    data_context_dir=f"/media/jenna/8TB-C/work/JUMP-download-images/JUMP-single-cell/0.download_data/data/images/{plate_id}/orig",
+    data_outline_context_dir=f"/media/jenna/8TB-C/work/JUMP-download-images/JUMP-single-cell/0.download_data/data/images/{plate_id}/outlines",
+    segmentation_file_regex=outline_to_orig_mapping,
+)[
+    [
+        "Nuclei_AreaShape_Compactness",
+        "Image_FileName_OrigDNA",
+    ]
+]
+
+
+print(poor_nuclei_outliers_cdf.shape)
+poor_nuclei_outliers_cdf.sort_values(by="Nuclei_AreaShape_Compactness", ascending=True).head(2)
+
+
+# ### Randomly sample outlier rows to visually inspect if the threshold looks to be optimized
+
+# In[10]:
+
+
+poor_nuclei_outliers_cdf.sample(n=5, random_state=0)
+
+
+# ### Save the original indices for failed single cells to compressed CSV files
+
+# In[11]:
+
+
+# Create a new dataframe for poor nuclei outliers
+poor_qc_indices = poor_nuclei_outliers[metadata_cols].copy()
+poor_qc_indices["original_index"] = poor_qc_indices.index  # Store original index
+poor_qc_indices = poor_qc_indices.reset_index(drop=True)  # Reset index
+poor_qc_indices["cqc.compactness_outlier"] = True
+
+# Merge both outlier dataframes on metadata columns and original_index
+failed_qc_indices = failed_qc_indices.merge(
+    poor_qc_indices,
+    on=[*metadata_cols, "original_index"],
+    how="outer"
+)
+
+# Fill missing values with False (ensuring only detected outliers are True)
+failed_qc_indices[["cqc.formfactor_displacement_outlier", "cqc.compactness_outlier"]] = (
+    failed_qc_indices[["cqc.formfactor_displacement_outlier", "cqc.compactness_outlier"]]
+    .fillna(False)
+)
+
+# Calculate percentage removed
+num_outliers = failed_qc_indices["original_index"].nunique()
+total_cells = len(df_merged_single_cells)
+percentage_removed = (num_outliers / total_cells) * 100
+
+# Sort by original index to maintain consistent order
+failed_qc_indices = failed_qc_indices.sort_values(by="original_index").reset_index(drop=True)
+
+# Save the indices for failed single-cells as a compressed CSV file
+failed_qc_indices.to_csv(Path(f"{qc_results_dir}/{plate_id}_failed_qc_indices.csv.gz"), compression="gzip")  # noqa: E501
```

Comment:
Consider using Parquet to help preserve data types and avoid ambiguous interpretations from CSV. These are compressed anyways, so we they'll appear like a Parquet file would. You could also take advantage of reading the entirety of the files as a single table more naturally, should you need it (a directory of Parquet files with the same schema may be treated as a single Parquet dataset).

Reply:
I did have this thought originally, but since parquet files (as far as I know) can not be viewed in a viewer (if that makes sense), I could not confirm what I was outputting. So, this was the easiest for me to confirm. Since I know that my outputs will look correct, then I will change these to be parquets instead.